cmd/proxy/main.go

```diff
@@ -0,0 +1,173 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"regexp"
+
+	"k8s.io/kube-openapi/pkg/common"
+	"k8s.io/kube-openapi/pkg/handler"
+	"k8s.io/kube-openapi/pkg/handler3"
+	"k8s.io/kube-openapi/pkg/spec3"
+	"k8s.io/kube-openapi/pkg/validation/spec"
+)
+
+func serve_v2(swagger spec.Swagger, h common.PathHandler) error {
+	o, err := handler.NewOpenAPIService(&swagger)
+	if err != nil {
+		return fmt.Errorf("failed to create OpenAPI v2 service: %v", err)
+	}
+	err = o.RegisterOpenAPIVersionedService("/openapi/v2", h)
+	if err != nil {
+		return fmt.Errorf("failed to register OpenAPI service: %v", err)
+	}
+	return nil
+}
+
+func serve_v3(v3 map[XKubernetesGroupVersion]spec3.OpenAPI, h common.PathHandler) error {
+	o, err := handler3.NewOpenAPIService(nil)
+	if err != nil {
+		return fmt.Errorf("failed to create OpenAPI v3 service: %v", err)
+	}
+
+	h.Handle("/openapi/v3", http.HandlerFunc(o.HandleDiscovery))
+	for gv := range v3 {
+		h.Handle("/openapi/v3/"+gv.Path(), http.HandlerFunc(o.HandleGroupVersion))
+		spec := v3[gv]
+		o.UpdateGroupVersion(gv.Path(), &spec)
+	}
+	return nil
+}
+
+func parse_v2(swagger_json []byte) (*spec.Swagger, error) {
+	swagger := spec.Swagger{}
+	if err := json.Unmarshal(swagger_json, &swagger); err != nil {
+		return nil, fmt.Errorf("failed to parse json: %v", err)
+	}
+	return &swagger, nil
+}
+
+type XKubernetesGroupVersion struct {
+	Group   string
+	Version string
+}
+
+func (gv XKubernetesGroupVersion) String() string {
+	if gv.Group == "" {
+		return gv.Version
+	}
+	return gv.Group + "/" + gv.Version
+}
+
+func (gv XKubernetesGroupVersion) Path() string {
+	if gv.Group == "" {
+		return "api/" + gv.Version
+	}
+	return "apis/" + gv.Group + "/" + gv.Version
+}
+
+var resourcePathRe = regexp.MustCompile(`^/api(s/(?P<group>[^/]+))?/(?P<version>[^/]+)/`)
+var groupIndexRe = resourcePathRe.SubexpIndex("group")
+var versionIndexRe = resourcePathRe.SubexpIndex("version")
+
+func parse_v3(openapi_json []byte) (XKubernetesGroupVersion, spec3.OpenAPI, error) {
+	openapi := spec3.OpenAPI{}
+	if err := json.Unmarshal(openapi_json, &openapi); err != nil {
+		return XKubernetesGroupVersion{}, spec3.OpenAPI{}, fmt.Errorf("failed to parse json: %v", err)
+	}
+	// Find the groupVersion.
+	if openapi.Paths == nil {
+		return XKubernetesGroupVersion{}, spec3.OpenAPI{}, fmt.Errorf("empty OpenAPI paths")
+	}
+	var gv *XKubernetesGroupVersion = nil
+	for path := range openapi.Paths.Paths {
+		matches := resourcePathRe.FindStringSubmatch(path)
```

Comment:
why use regex instead of directly searching for a specific path?

Reply:
For the first occurence, we need the regex, after that we could be searching for the path, sure.

Reply:
To be fair also, I'm not planning on keeping this code super long :-)